File: drivers/rust/src/redwire/frame.rs
//! Wire frame layout — mirrors the engine's
//! `reddb::wire::redwire::frame`. Kept duplicated rather than
//! shared via a crate dependency so the driver can build without
//! the engine.

pub const FRAME_HEADER_SIZE: usize = 16;
pub const MAX_FRAME_SIZE: u32 = 16 * 1024 * 1024;

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Frame {
    pub kind: MessageKind,
    pub flags: Flags,
    pub stream_id: u16,
    pub correlation_id: u64,
    pub payload: Vec<u8>,
}

impl Frame {
    pub fn new(kind: MessageKind, correlation_id: u64, payload: Vec<u8>) -> Self {
        Self {
            kind,
            flags: Flags::empty(),
            stream_id: 0,
            correlation_id,
            payload,
        }
    }

    pub fn encoded_len(&self) -> u32 {
        (FRAME_HEADER_SIZE + self.payload.len()) as u32
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
#[repr(u8)]
pub enum MessageKind {
    Query = 0x01,
    Result = 0x02,
    Error = 0x03,
    BulkInsert = 0x04,
    BulkOk = 0x05,
    BulkInsertBinary = 0x06,
    QueryBinary = 0x07,
    BulkInsertPrevalidated = 0x08,
    Hello = 0x10,
    HelloAck = 0x11,
    AuthRequest = 0x12,
    AuthResponse = 0x13,
    AuthOk = 0x14,
    AuthFail = 0x15,
    Bye = 0x16,
    Ping = 0x17,
    Pong = 0x18,
    Get = 0x19,
    Delete = 0x1A,
    DeleteOk = 0x1B,
}

impl MessageKind {
    pub fn from_u8(byte: u8) -> Option<Self> {
        match byte {
            0x01 => Some(Self::Query),
            0x02 => Some(Self::Result),
            0x03 => Some(Self::Error),
            0x04 => Some(Self::BulkInsert),
            0x05 => Some(Self::BulkOk),
            0x06 => Some(Self::BulkInsertBinary),
            0x07 => Some(Self::QueryBinary),
            0x08 => Some(Self::BulkInsertPrevalidated),
            0x10 => Some(Self::Hello),
            0x11 => Some(Self::HelloAck),
            0x12 => Some(Self::AuthRequest),
            0x13 => Some(Self::AuthResponse),
            0x14 => Some(Self::AuthOk),
            0x15 => Some(Self::AuthFail),
            0x16 => Some(Self::Bye),
            0x17 => Some(Self::Ping),
            0x18 => Some(Self::Pong),
            0x19 => Some(Self::Get),
            0x1A => Some(Self::Delete),
            0x1B => Some(Self::DeleteOk),
            _ => None,
        }
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Flags(u8);

impl Flags {
    pub const COMPRESSED: Self = Self(0b0000_0001);
    pub const MORE_FRAMES: Self = Self(0b0000_0010);

    pub const fn empty() -> Self {
        Self(0)
    }

    pub const fn bits(self) -> u8 {
        self.0
    }

    pub const fn from_bits(bits: u8) -> Self {
        Self(bits)
    }
}

File: drivers/zig/src/redwire/frame.zig
// RedWire frame layout. Mirrors `src/wire/redwire/frame.rs` and
// the Rust driver's `drivers/rust/src/redwire/frame.rs`. Keep the
// byte-level constants in lockstep — the wire is a public ABI.

const std = @import("std");

pub const FRAME_HEADER_SIZE: usize = 16;
pub const MAX_FRAME_SIZE: u32 = 16 * 1024 * 1024;
pub const KNOWN_FLAGS: u8 = 0b0000_0011;

pub const Flags = struct {
    pub const COMPRESSED: u8 = 0b0000_0001;
    pub const MORE_FRAMES: u8 = 0b0000_0010;
};

pub const MessageKind = enum(u8) {
    query = 0x01,
    result = 0x02,
    err = 0x03,
    bulk_insert = 0x04,
    bulk_ok = 0x05,
    bulk_insert_binary = 0x06,
    query_binary = 0x07,
    bulk_insert_prevalidated = 0x08,
    hello = 0x10,
    hello_ack = 0x11,
    auth_request = 0x12,
    auth_response = 0x13,
    auth_ok = 0x14,
    auth_fail = 0x15,
    bye = 0x16,
    ping = 0x17,
    pong = 0x18,
    get = 0x19,
    delete = 0x1A,
    delete_ok = 0x1B,

    pub fn fromByte(b: u8) ?MessageKind {
        return switch (b) {
            0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08 => @enumFromInt(b),
            0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1A, 0x1B => @enumFromInt(b),
            else => null,
        };
    }
};

/// Header laid out in wire order. Encoded little-endian everywhere.
/// Kept as a plain struct because Zig's `extern struct` field
/// alignment for `u64` would force a 4-byte gap between `stream_id`
/// and `correlation_id`; we encode/decode by hand instead.
pub const FrameHeader = struct {
    length: u32,
    kind: u8,
    flags: u8,
    stream_id: u16,
    correlation_id: u64,
};

pub const Frame = struct {
    kind: MessageKind,
    flags: u8 = 0,
    stream_id: u16 = 0,
    correlation_id: u64,
    payload: []const u8,

    pub fn init(kind: MessageKind, correlation_id: u64, payload: []const u8) Frame {
        return .{
            .kind = kind,
            .flags = 0,
            .stream_id = 0,
            .correlation_id = correlation_id,
            .payload = payload,
        };
    }

    pub fn encodedLen(self: Frame) u32 {
        return @intCast(FRAME_HEADER_SIZE + self.payload.len);
    }
};

/// Write a frame header into `dst`. Caller guarantees `dst.len ==
/// FRAME_HEADER_SIZE` and that `total_length` already accounts for
/// any compression done on the payload.
pub fn writeHeader(
    dst: []u8,
    total_length: u32,
    kind: MessageKind,
    flags: u8,
    stream_id: u16,
    correlation_id: u64,
) void {
    std.debug.assert(dst.len >= FRAME_HEADER_SIZE);
    std.mem.writeInt(u32, dst[0..4], total_length, .little);
    dst[4] = @intFromEnum(kind);
    dst[5] = flags;
    std.mem.writeInt(u16, dst[6..8], stream_id, .little);
    std.mem.writeInt(u64, dst[8..16], correlation_id, .little);
}

/// Decode just the header. Doesn't validate — call `validateHeader`
/// after to enforce length / flag bounds.
pub fn readHeader(src: []const u8) !FrameHeader {
    if (src.len < FRAME_HEADER_SIZE) return error.FrameTruncated;
    return FrameHeader{
        .length = std.mem.readInt(u32, src[0..4], .little),
        .kind = src[4],
        .flags = src[5],
        .stream_id = std.mem.readInt(u16, src[6..8], .little),
        .correlation_id = std.mem.readInt(u64, src[8..16], .little),
    };
}

pub fn validateHeader(h: FrameHeader) !void {
    if (h.length < FRAME_HEADER_SIZE or h.length > MAX_FRAME_SIZE) {
        return error.FrameInvalidLength;
    }
    if ((h.flags & ~KNOWN_FLAGS) != 0) {
        return error.FrameUnknownFlags;
    }
}
